docs/modules/ROOT/pages/usage.adoc

```diff
@@ -0,0 +1,390 @@
+
+=== Storage and events
+
+The component's storage and events must be added to the contract's `Storage` struct and `Event` enum respectively.
+If the component doesn't define any events, the compiler will still create an empty event enum inside the component module.
+
+[,javascript]
+----
+#[starknet::contract]
+mod MyContract {
+    use openzeppelin::security::InitializableComponent;
+
+    component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);
+
+    #[storage]
+    struct Storage {
+        #[substorage(v0)]
+        initializable: InitializableComponent::Storage
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        InitializableEvent: InitializableComponent::Event
+    }
+
+    (...)
+}
+----
+
+The `#[substorage(v0)]` attribute must be included for each component in the `Storage` trait.
+This allows the contract to have indirect access to the component's storage.
+
+The `#[flat]` attribute for events in the `Event` enum, however, is not required.
+Component events are not flattened in the component itself because it would remove the event ID from the event log.
+Note that if contracts do not flatten component events, the first key in the event log will be the component ID.
+By flattening the component event in the contract, the first key will be the event ID.
+
+=== Implementations
+
+:accessing-storage: xref:/usage.adoc#accessing_component_storage[Accessing component storage]
+
+Components come with granular implementations of different interfaces.
+This allows contracts to integrate only the implementations that they'll use and avoid unnecessary bloat.
+Integrating an implementation looks like this:
+
+[,javascript]
+----
+mod MyContract {
+    use openzeppelin::security::InitializableComponent;
+
+    component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);
+
+    (...)
+
+    // Gives the contract access to the implementation methods
+    impl InitializableImpl =
+        InitializableComponent::InitializableImpl<ContractState>;
+}
+----
+
+Defining an `impl` gives the contract access to the methods within the implementation from the component.
+For example, `is_initialized` is defined in the `InitializableImpl`.
+A function on the contract level can expose it like this:
+
+[,javascript]
+----
+#[starknet::contract]
+mod MyContract {
+    use openzeppelin::security::InitializableComponent;
+
+    component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);
+
+    (...)
+
+    impl InitializableImpl =
+        InitializableComponent::InitializableImpl<ContractState>;
+
+    #[external(v0)]
+    fn is_initialized(ref self: ContractState) -> bool {
+        self.initializable.is_initialized()
+    }
+}
+----
+
+Notice that the function must pass the state (`self`) and the component's storage (`initializable`) before finally accessing the implementation method.
```

Comment:
I'm not sure what this sentence intends to point out. How to call a method from an impl on the self.storage member? what do you mean by "passing ... the component's storage"?

Reply:
> I'm not sure what this sentence intends to point out

It's attempting to show the syntax for calling component implementation methods. Maybe there's better names to use when describing it?

> what do you mean by "passing ... the component's storage"?

It's referring to the `storage` arg in the `component!` macro and the member of the `Storage` struct:

```
component!(path: InitializableComponent, storage: initializable, event: InitializableEvent);

#[storage]
struct Storage {
    #[substorage(v0)]
    initializable: InitializableComponent::Storage
}
```

I'll try to rework this and come up with better ways to describe the process